File: .git-rewrite/t/experiments/results/reports/exp_kg2a1_b0/code/shortener.py
import secrets
import string
from db import code_exists

ALPHABET = string.ascii_letters + string.digits
CODE_LENGTH = 7
MAX_ATTEMPTS = 8


def _generate_code():
    return "".join(secrets.choice(ALPHABET) for _ in range(CODE_LENGTH))


def generate_unique_code(db_path):
    for _ in range(MAX_ATTEMPTS):
        code = _generate_code()
        if not code_exists(db_path, code):
            return code
    raise RuntimeError("Failed to generate a unique short code after max attempts")

File: .git-rewrite/t/experiments/results/reports/exp_d6lkz5pg/code/src/shortener.ts
import { customAlphabet } from "nanoid";
import { insertUrl, getByCode } from "./db.js";

const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
const CODE_LENGTH = 7;
const MAX_RETRIES = 5;

const generate = customAlphabet(ALPHABET, CODE_LENGTH);

export function generateCode(): string {
  return generate();
}

export function shorten(original_url: string, expiresInSeconds: number | null): string {
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    const code = generateCode();
    const existing = getByCode(code);
    if (existing) continue;

    const expires_at = expiresInSeconds ? Math.floor(Date.now() / 1000) + expiresInSeconds : null;
    insertUrl(code, original_url, expires_at);
    return code;
  }
  throw new Error("Failed to generate a unique short code after max retries");
}
